### PR TITLE
fx quant: fix bug with fusion on a module used by multiple nodes

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68178

Summary:

Before this PR, the FX quantization fusion logic incorrectly fused
modules which had multiple nodes using them.  For example,

```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(1, 1, 1)
        self.bn = torch.nn.BatchNorm2d(1)

    def forward(self, x):
        x = self.conv(x)
        x = self.conv(x)
        x = self.bn(x)
        return x
```

the second conv-bn would be fused, which would change the meaning of
the first call to the conv.

This PR changes the pattern matching logic to disable matching
for any nodes which are not the last node in a pattern and which
refer to modules which have multiple module_uses.

Note that this definition of "module_use" is different from the "use" in the
FX framework.  In FX, a use is referring to which nodes use the output
of a node.  "module_use" refers to how many FX nodes use a module
instance.

Test Plan:

```
python test/test_quantization.py TestFuseFx.test_fusion_turns_off_for_multi_use_intermediate_nodes
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32357549](https://our.internmc.facebook.com/intern/diff/D32357549)

### CODE REVIEW

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32357549).

File: torch/ao/quantization/fx/_lower_to_native_backend.py [hunk 1]
```diff
@@ -2,23 +2,29 @@
...
         MatchAllNode, MatchAllNode, MatchAllNode)
     modules = dict(model.named_modules())
     nodes = list(model.graph.nodes)
     # TODO: maybe orgnize this better (e.g. break down to more functions)
     # to make this function more readable
+    module_name_to_num_node_users = \
+        calculate_module_name_to_num_node_users(model.graph)
     for n in model.graph.nodes:
-        if not is_match(modules, n, pattern):
+        if not is_match(modules, n, pattern, module_name_to_num_node_users):
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_lower_to_native_backend.py:27]: does it work if we just set `max_uses = 1` here?

Comment (vkuzo) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:27]: sorry, not sure I understand the question yet, could you provide some more context?

`max_uses` is an FX concept, this PR is adding checking for "number of nodes which use this module instance" which is not a concept that exists in FX.

Comment (jerryzh168) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:27]: oh right, sorry I mixed the use of Node and use of module instance

File: torch/ao/quantization/fx/match_utils.py [hunk 2]
```diff
@@ -49,16 +75,30 @@ def is_match(modules, node, pattern, max_uses=sys.maxsize):
     if len(node.users) > max_uses:
         return False
 
+    if (
+        # This is True if we are not at the last node in the pattern
+        max_uses < sys.maxsize and
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/match_utils.py:80]: could you explain a little bit more about what does this mean and why do we need this check?

**Review by jerryzh168 — APPROVED** [review]

File: torch/ao/quantization/fx/match_utils.py (hunk 2 above)
Comment (vkuzo) [reply torch/ao/quantization/fx/match_utils.py:80]: you are right, this check is incorrect, if the last node in the pattern is using a module which is used by other nodes we also cannot fuse without further analysis.  Will remove this.

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32357549).

**Comment by vkuzo** [comment]
@jerryzh168 , please check out if the modified test cases are what you expect

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32357549).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32357549).

File: test/quantization/fx/test_quantize_fx.py [hunk 3]
```diff
@@ -445,6 +446,53 @@ def forward(self, x):
         })
         self.checkGraphModuleNodes(m, expected_node=ns.call_module(MyConvReLU))
 
+    def test_fusion_turns_off_for_multi_use_intermediate_nodes(self):
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:449]: can we also have a test for multiple use of relu module? like conv - relu?

**Review by jerryzh168 — APPROVED** [review]
lg, one comment inline for test

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32357549).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32357549).

File: torch/ao/quantization/fx/match_utils.py [hunk 4]
```diff
@@ -22,17 +23,49 @@
 MatchResult = Tuple[Node, List[Node], Optional[Pattern], QuantizeHandler,
                     QConfigAny]
 
+# Module types in this list are allowed to be used by multiple 'call_module'
+# nodes. Module types not in this list will not be matched if they are used by
+# multiple 'call_module' nodes.
+MODULE_TYPES_SAFE_TO_USE_BY_MULTIPLE_NODES = set([
+    torch.nn.ReLU,
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/match_utils.py:30]: I remember Kimish reported another case, there is a Lienar - relu pattern, and then then people access the bias in linear, should we add Linear/Conv here as well?

Comment (jerryzh168) [reply torch/ao/quantization/fx/match_utils.py:30]: Oh sorry, this is safe to use, sounds good

File: test/quantization/fx/test_quantize_fx.py [hunk 5]
```diff
@@ -469,6 +470,105 @@ def forward(self, x):
...
+        ]
+        expected_occurrence = {
+            ns.call_module(nni.ConvBn2d): 0,
+        }
+        self.checkGraphModuleNodes(
+            m,
+            expected_node_list=expected_nodes,
+            expected_node_occurrence=expected_occurrence)
+
+    def test_fusion_works_for_multi_use_intermediate_relu_nodes(self):
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:519]: there is one more test case, reported by Kimish, could you add that as well (probably need to change the matching code to address this as well), something like

```
class M(torch.nn.Module):
    def forward(self, x):
           x = self.linear(x)
           bias = self.linear.bias
           ...
```
I can take it as well if you don't have time

**Pull request closed without merging**